rs/bitcoin/adapter/src/lib.rs

```diff
@@ -154,30 +158,64 @@ pub struct AdapterState {
     last_received_at: Arc<RwLock<Option<Instant>>>,
     /// The field contains how long the adapter should wait to before becoming idle.
     idle_seconds: u64,
+
+    awake_tx: watch::Sender<()>,
 }
 
 impl AdapterState {
     /// Crates new instance of the AdapterState.
     pub fn new(idle_seconds: u64) -> Self {
-        Self {
+        let (awake_tx, _) = watch::channel(());
+        let state = Self {
             last_received_at: Arc::new(RwLock::new(None)),
             idle_seconds,
-        }
-    }
-
-    /// Returns if the adapter is idle.
-    pub fn is_idle(&self) -> bool {
-        match *self.last_received_at.read() {
-            Some(last) => last.elapsed().as_secs() > self.idle_seconds,
-            // Nothing received yet still in idle from startup.
-            None => true,
-        }
+            awake_tx,
+        };
+        state
     }
 
     /// Updates the current state of the adapter given a request was received.
     pub fn received_now(&self) {
         // Instant::now() is monotonically nondecreasing clock.
         *self.last_received_at.write() = Some(Instant::now());
+        // TODO: perhaps log something if this fails
+        let _ = self.awake_tx.send(());
+    }
+
+    /// A future that returns when/if the adapter becomes/is awake.
```

Comment:
nit: I think active is a better wording here. Then you can also rename the function to 

`pub fn async active(..);
`

Reply:
done